cli/lib/kontena/cli/stacks/upgrade_command.rb

```diff
@@ -81,91 +81,109 @@ def process_data(old_data, new_data)
       logger.debug { "Master stacks: #{old_data.keys.join(",")} YAML stacks: #{new_data.keys.join(",")}" }
 
       new_data.reverse_each do |stackname, data|
-        reader = data[:loader].reader
-        set_env_variables(stackname, current_grid) # set envs for execution time
-        data[:stack_data] = reader.execute(
-          values: data[:variables],
-          defaults: old_data[stackname].nil? ? nil : old_data[stackname][:stack_data]['variables'],
-          parent_name: data[:parent_name],
-          name: data[:name]
-        )
+        spinner "Processing stack #{pastel.cyan(stackname)}"
+        process_stack_data(stackname, data, old_data)
         hint_on_validation_notifications(reader.notifications, reader.loader.source)
         abort_on_validation_errors(reader.errors, reader.loader.source)
       end
 
-      set_env_variables(stack_name, current_grid) # restore envs
-
+      old_set = Kontena::Stacks::StackDataSet.new(old_data)
+      new_set = Kontena::Stacks::StackDataSet.new(new_data)
+      if skip_dependencies?
+        [old_set, new_set].each(&:remove_dependencies)
```

Comment:
Huh... wouldn't it be easier to just change `gather_master_data` to not recurse instead, and then use some variant of `loader.to_h` instead of `loader.flat_dependencies`? It looks like the later might need some slight refactoring of `flat_dependencies` to split out the `{ basename => { ...., variables: ... } }` logic, but that's what this PR is for...?

Loading everything recursively only to just strip them out again does not really count as "refactoring" to me ^___^

Reply:
We need to load everything recursively so that variable referencing works.

Reply:
By variable referencing, do you mean the way that [`StackFileLoader#flat_dependencies`](https://github.com/kontena/kontena/blob/c5fecb2b9370f2742285a0f1bf7842b01b9cef62/cli/lib/kontena/cli/stacks/yaml/stack_file_loader.rb#L114-L119) manipulates the `values_from_options` to handle args like `-v deptest-foo.url=bar`?

I figure the `flat_dependencies` method just needs to refactored to split out the initial non-recursive part, and then just use that instead of `flat_dependencies` if `--skip-dependencies`:

```
      def flat_parent(basename, opts = {})
        opt_variables = opts[:variables] || {}

        return {
          basename => self.to_h.merge(opts).merge(
            name: basename,
            variables: opt_variables.reject { |k, _| k.include?('.') }
          )
        }
      end
```

I'm not aware of any way to reference the child stack variables from the parent?

Reply:
> I'm not aware of any way to reference the child stack variables from the parent?

It's not yet possible, but it should be in the future.. see #2856

Reply:
Fair enough. Should this somehow be passed as a flag to the `ChangeResolver` instead, so that it only returns the top-level stack in `upgraded_stacks`?

Reply:
I had it first in the `ChangeResolver` but somehow it felt weird. I think we can improve this in another pr (easier to review smaller pr's)?